src/Statistics.jl

```diff
@@ -0,0 +1,135 @@
+uses a cyclic proximal point scheme. Optionally one can provide
+weights $w_i$ for the weighted Riemannian median. The general formula to compute
+the minimizer reads
+````math
+\argmin_{y\in\mathcal M}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemannian [`distance`](@ref).
+
+The cyclic proximal point can be run with a random cyclic order by setting
+`use_rand` to `true`
+
+Optionally you can provide `x0`, the starting point (by default set to the first
+data point). `stop_iter` denotes the maximal number of iterations to perform and
+the `kwargs...` are passed to [`isapprox`](@ref) to stop, when the inimal change
+between two iterates is small. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in Algorithm 4.3 and 4.4 in 
+> Bačák, M: Computing Medians and Means in Hadamard Spaces.
+> SIAM Journal on Optimization (2014), 24(3), pp. 1542–1566,
+> doi: [10.1137/140953393](https://doi.org/10.1137/140953393),
+> arxiv: [1210.2145](https://arxiv.org/abs/1210.2145)
+"""
+function median(M::Manifold, x::AbstractVector, w::AbstractWeights = weights(ones(length(x)) / length(x)); x0=x[1], kwargs...)
+    y = similar_result(M, median, x[1])
+    copyto!(y,x0)
+    return median!(M, y, x; kwargs...)
+end
+@doc doc"""
+    median!(M,y,x,w)
+
+computes the [`median`](@ref) in-place in `y` where the initial value of `y` is
+the starting point of the algorithm.
+"""
+function median!(M::Manifold, y, x::AbstractVector, w::AbstractWeights = weights(ones(length(x)) / length(x));
+            stop_iter=100000,
+            use_random = false,
+            kwargs...
+        ) where {T}
+    n = length(x)
+    yold = similar_result(M,median,y)
+    copyto!(yold,y)
+    order = collect(1:n)
+    v = zero_tangent_vector(M,y)
+    for i=1:stop_iter
+        λ = n/(2*i)
+        copyto!(yold,y)
+        use_random && shuffle!(order)
+        for j=1:n
+            t = min( λ * w[order[j]] / distance(M,y,x[order[j]]) , 1 )
+            log!(M, v, y, x[order[j]])
+            y = exp(M, y, t* v)
+        end
+        isapprox(M, y, yold; kwargs...) && break
+    end
+    return y
+end
+median(M::Euclidean, x::AbstractVector, w::AbstractWeights) = median(x,w)
+median(M::Euclidean, x::AbstractVector) = median(x)
+@doc doc"""
+    var(M,x, corrected::Bool=true, mean=mean(M,x))
+
+compute the variance of the points in the  vector `x` of length `n`
+on the [`Manifold`](@ref) `M`, i.e.
+````
+    \frac{1}{n'}\sum_{i=1}^n d_{\mathcal M}^2(m,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemannian [`distance`](@ref)
+and $n'=n-1$ if `corrected=true` and $n'=n$ otherwise.
+
+A mean might be provided and is otherwise computed.
+"""
+var(M::Manifold, x::AbstractVector, corrected::Bool=true, mean=mean(M,x)) = sum( distance.(Ref(M), Ref(mean), x).^2 )/(length(x)-Int(corrected))
```

Comment:
Can you add the weighted case as well?

Reply:
Add for `Euclidean` the default `var`

Reply:
Do you have a formula for the weighted one? Especially, I am unsure whether it's weights or weights-squared in the formula (i.e. `(weights.*distance(Ref(M),Ref(mean),x)).^2` or `weights.*(distance(Ref(M),Ref(mean),x)).^2` within the sum. How does the correction term look like then?

Reply:
See the StatsBase implementation of `_moment2`: https://github.com/JuliaStats/StatsBase.jl/blob/master/src/moments.jl#L234-L244. Here's an (untested) implementation of these functions based on theirs:

```julia
function var(
    M::Manifold,
    x::AbstractVector,
    w::AbstractWeights,
    mean = mean(M, x, w);
    corrected::Bool = false,  # for var corrected defaults to false
)
    wv = convert(Vector, w)
    sqdist = wv .* distance.(Ref(M), Ref(mean), x) .^ 2
    s = sum(sqdist)
    c = w isa Weights ? varcorrection(length(x), corrected) / w.sum :
        varcorrection(w, corrected)
    return c * s
end

function var(M::Manifold, x::AbstractVector, mean = mean(M, x, w); kwargs...)
    return var(M, x, weights(ones(length(x))), mean; kwargs...)
end

function std(
    M::Manifold,
    x::AbstractVector,
    w::AbstractWeights,
    mean = mean(M, x, w);
    corrected::Bool = true,
)
    return sqrt(var(M, x, w; corrected = corrected))
end

function std(M::Manifold, x::AbstractVector, mean = mean(M, x, w); kwargs...)
    return std(M, x, weights(ones(length(x))), mean; kwargs...)
end

function mean_and_var(
    M::Manifold,
    x::AbstractVector,
    w::AbstractWeights = weights(ones(length(x)));
    kwargs...,
)
    m = mean(M, x, w)
    v = var(M, x, w, m; kwargs...)
    return m, v
end

function mean_and_std(
    M::Manifold,
    x::AbstractVector,
    w::AbstractWeights = weights(ones(length(x)));
    kwargs...,
)
    m = mean(M, x, w)
    s = std(M, x, w, m; kwargs...)
    return s, v
end
```

Reply:
I made some fixes to the above snippet

Reply:
Hm, this has a problem with the default (non keyword) arguments, since one case of var occurs double (manifold, array, mean), which occurs in both cases. I think for the second var, the mean can not have a default.